packages/luma/src/luma-image-model.ts

```diff
@@ -0,0 +1,274 @@
+        ),
+      });
+
+      switch (statusResponse.state) {
+        case 'completed':
+          if (!statusResponse.assets?.image) {
+            throw new InvalidResponseDataError({
+              data: statusResponse,
+              message: `Image generation completed but no image was found.`,
+            });
+          }
+          return statusResponse.assets.image;
+        case 'failed':
+          throw new InvalidResponseDataError({
+            data: statusResponse,
+            message: `Image generation failed.`,
+          });
+        case 'dreaming':
+        case 'queued':
+          break;
+      }
+
+      attemptCount++;
+      await new Promise(resolve =>
+        setTimeout(resolve, this.pollIntervalMillis),
+      );
+    }
+
+    throw new Error(
+      `Image generation timed out after ${this.maxPollAttempts} attempts.`,
+    );
+  }
+
+  private createLumaErrorHandler() {
+    return createJsonErrorResponseHandler({
+      errorSchema: lumaErrorSchema,
+      errorToMessage: (error: LumaErrorData) =>
+        error.detail[0].msg || 'Unknown error',
+    });
+  }
+
+  private getLumaGenerationsUrl(generationId?: string) {
+    return `${this.config.baseURL}/dream-machine/v1/generations/${
+      generationId || 'image'
+    }`;
+  }
+
+  private async downloadImage(
+    url: string,
+    abortSignal: AbortSignal | undefined,
+  ): Promise<Uint8Array> {
+    const { value: buffer } = await getFromApi({
+      url,
+      // No specific headers should be needed for this request as it's a
+      // generated image provided by Luma.
+      abortSignal,
+      failedResponseHandler: createStatusCodeErrorResponseHandler(),
+      successfulResponseHandler: createBinaryResponseHandler(),
+      fetch: this.config.fetch,
+    });
+    return new Uint8Array(buffer);
+  }
+}
+
+const lumaRequestSchema = z.object({
+  generation_type: z.literal('image'),
+  model: z.string(),
+  prompt: z.string(),
+  aspect_ratio: z.string().nullish(),
+  callback_url: z.string().nullish(),
+  image_ref: z
+    .array(
+      z.object({
+        url: z.string(),
+        weight: z.number(),
+      }),
+    )
+    .nullish(),
+  style_ref: z
+    .array(
+      z.object({
+        url: z.string(),
+        weight: z.number(),
+      }),
+    )
+    .nullish(),
+  character_ref: z
+    .object({
+      identity0: z.object({
+        images: z.array(z.string()),
+      }),
+    })
+    .nullish(),
+  modify_image_ref: z
+    .object({
+      url: z.string(),
+      weight: z.number(),
+    })
+    .nullish(),
+});
+
+const lumaGenerationResponseSchema = z.object({
+  id: z.string(),
+  generation_type: z.literal('image'),
+  state: z.enum(['queued', 'dreaming', 'completed', 'failed']),
+  failure_reason: z.string().nullish(),
+  created_at: z.string(),
+  assets: z
+    .object({
+      image: z.string(), // URL of the generated image
+    })
+    .nullish(),
+  model: z.string(),
+  request: lumaRequestSchema,
+});
```

Comment:
is all of this needed or can it be slimmed down?

Reply:
reduced quite a bit